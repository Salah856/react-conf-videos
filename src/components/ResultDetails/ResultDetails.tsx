import * as React from 'react';

import { compose, flatten, pluck, values } from 'ramda';
import { mapProps } from 'recompose';

import { IndexedConferences } from '../../domain';

import styles from './ResultDetails.scss';

type Props = { conferences: IndexedConferences };
type MappedProps = {
  countVids: number;
  countConfs: number;
  countVidsS: string;
  countConfsS: string;
};

const getAllVideos = compose<any, any, any, string[]>(
  flatten,
  values,
  pluck('videos')
);

export const computeResultDetails = (conferences: IndexedConferences) => {
  const confKeys = Object.keys(conferences);
  const videos = getAllVideos(conferences);
  return {
    countVids: videos.length,
    countVidsS: videos.length === 1 ? 'video' : 'videos',
    countConfs: confKeys.length,
    countConfsS: confKeys.length === 1 ? 'conference' : 'conferences'
  };
};

const ResultDetailsInner: React.FunctionComponent<MappedProps> = ({
  countVids,
  countConfs,
  countVidsS,
  countConfsS
}) => (
  <p className={styles.resultsCount}>
    <span className={styles.resultsNumber}> {countVids} </span> {countVidsS}
    <span className={styles.resultsNumber}> {countConfs} </span> {countConfsS}
  </p>
);

export const ResultDetails = mapProps<MappedProps, Props>(({ conferences }) => {
  return {
    ...computeResultDetails(conferences)
  };
})(ResultDetailsInner);
