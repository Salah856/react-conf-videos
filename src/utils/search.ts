import { compose, toLower, trim } from 'ramda'
import { remove as removeDiacritics } from 'diacritics'

export const cleanQuery = compose(
  removeDiacritics,
  toLower,
  trim
)

export const isFilterEmpty = (filterValue: string = '') => filterValue.trim() === ''
