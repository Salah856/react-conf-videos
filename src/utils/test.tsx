import { mount, shallow } from 'enzyme';
import * as React from 'react';
import { ReactElement } from 'react';
import { MemoryRouter } from 'react-router';
import { StateObservable } from 'redux-observable';
import { createMockStore } from 'redux-test-utils';
import { Subject } from 'rxjs';
import * as Immutable from 'seamless-immutable';

import { ApplicationState } from 'redux/modules';
import {
  Conference,
  IndexedConferences,
  IndexedPresenters,
  IndexedVideos,
  Presenter,
  Video
} from '../domain';

// cast to ApplicationState so TS doesn't complain when sending in partials to Epics for testing
export const mockStore = (state = {}) =>
  new StateObservable<ApplicationState>(
    new Subject(),
    state as ApplicationState
  );

const isStore = (store: any) => store.hasOwnProperty('isActionTypeDispatched');

// shallow mount component with a store in context
export const shallowWithStore = (storeP: any, component: ReactElement<{}>) => {
  const store = isStore(storeP) ? storeP : createMockStore(storeP);
  return shallow<any, any>(component, { context: { store } });
};

export const mountWithStore = (storeP: any, component: ReactElement<{}>) => {
  const store = isStore(storeP) ? storeP : createMockStore(storeP);
  return mount<any, any>(component, { context: { store } });
};

export const wrapWithMemoryRouter = (Component: React.ReactElement<any>) => (
  // key within location prop keeps changing, so manually set
  // so snapshorts are consistent
  <MemoryRouter initialEntries={[{ key: 'xxxx' }]}>{Component}</MemoryRouter>
);

/** Mock data */
export const mockVideo = (title = 'test title'): Video => ({
  link: 'a link',
  split: '',
  title: title || 'test title',
  length: '12:34',
  lightning: false,
  presenter: 'siomn carter',
  embeddableLink: 'a link'
});

export const mockVideos = (): IndexedVideos => ({
  aaa: mockVideo('aaa'),
  bbb: mockVideo('bbb'),
  ccc: mockVideo('ccc')
});

export const mockPresenter = (name = 'simon carter'): Presenter => ({
  name
});

export const mockPresenters = (): IndexedPresenters => ({
  xxx: mockPresenter(),
  zzz: mockPresenter('zzz')
});

export const mockConference = (): Conference => ({
  date: 'XX/YY/ZZZ',
  title: 'fake conf title',
  website: 'fake url',
  playlist: 'day 1',
  videos: ['aaa', 'bbb', 'ccc']
});

export const mockIndexedConferences = (): IndexedConferences => ({
  ['XXXX']: mockConference()
});

export const mockRouterState = (
  pathname = '/conference/8424e37df85b9eccbe48e9a55d93845e/react-conf-2018'
) => ({
  router: {
    location: {
      pathname,
      search: '',
      hash: ''
    }
  }
});

export const mockConferencePageSlice = () => ({
  conferencePage: {
    selectedConferenceId: 'XXXX',
    conference: mockConference()
  }
});

export const mockDataState = () =>
  Immutable({
    data: {
      conferences: mockIndexedConferences(),
      presenters: mockPresenters(),
      videos: mockVideos()
    }
  });
