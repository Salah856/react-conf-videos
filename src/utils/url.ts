import { compose, replace, toLower } from 'ramda';

// replaces white space with - and lower case
export const sluggifyUrl = compose(
  replace(/\s/g, '-'),
  toLower
);
