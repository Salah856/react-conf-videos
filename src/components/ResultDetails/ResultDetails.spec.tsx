import * as React from 'react'
import toJSON from 'enzyme-to-json'
import { mount } from 'enzyme'

import { ResultDetails } from './ResultDetails'
import { mockIndexedConferences } from 'utils'

describe('ResultDetails', () => {
  it('should render', () => {
    // arrange
    const props = {
      conferences: mockIndexedConferences()
    }

    // act
    const comp = mount(<ResultDetails {...props} />)

    // assert
    expect(toJSON(comp)).toMatchSnapshot()
  })
})
