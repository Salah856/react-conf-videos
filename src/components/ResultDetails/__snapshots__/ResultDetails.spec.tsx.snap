// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultDetails should render 1`] = `
<Component
  conferences={
    Object {
      "XXXX": Object {
        "date": "XX/YY/ZZZ",
        "playlist": "day 1",
        "title": "react conf 2018",
        "videos": Array [
          "aaa",
          "bbb",
          "ccc",
        ],
        "website": "fake url",
      },
    }
  }
>
  <p
    className="resultsCount"
  >
    <strong>
       
      3
       
    </strong>
     
    videos
    <strong>
       
      1
       
    </strong>
     
    conference
  </p>
</Component>
`;
